### DISCUSSION TITLE
Fastapi db local session example

### DISCUSSION THREAD

**Discussion by Faridalim** [piccolo-orm/piccolo#131] [Q&A]
Hi, thanks for sharing this ORM. It looks simple and easy which is a good thing.

Could anybody give me an example of using Piccolo in Fastapi especially when using database local session. I mean, how to use Piccolo if i have many files in Fastapi that need db connections? How to use it with connection pool? Thank you very much

**Comment by dantownsend** [comment]
@Faridalim The best thing to try is `piccolo asgi new` on the command line, which creates an example FastAPI app which uses Piccolo.

To setup a connection pool, you have something like this in your app file:

```python
app = FastAPI()


@app.on_event("startup")
async def open_database_connection_pool():
    try:
        engine = engine_finder()
        await engine.start_connection_pool()
    except Exception:
        print("Unable to connect to the database")


@app.on_event("shutdown")
async def close_database_connection_pool():
    try:
        engine = engine_finder()
        await engine.close_connection_pool()
    except Exception:
        print("Unable to connect to the database")
```


You shouldn't have to pass the sessions around with Piccolo. It does have [transaction](https://piccolo-orm.readthedocs.io/en/latest/piccolo/query_types/transactions.html) support though if you need it.

Does that help?
